imgurbash2

```diff
@@ -22,216 +22,599 @@
+        echo "    ERR [$JOB_ID] $msg" >> "$LOG_FILE"
+    fi
+}
+
 
-	# loop through arguments
-	for file do
-		# upload the image
-		response=$(upload_image "$file")
-		url=""
+# Convenience method for err()
+fail() {
+    err "$@"
+    exit 1
+}
 
-		# get image url link
-		grep -qo '"status":200' <<< "$response"  &&  url=$(get_value_from_response "$response" "link" | sed -e 's/\\//g')
 
-		if [ -z "$url" ]; then
-			echo "Upload failed for $file:  "$(get_value_from_response "$response" "error") >&2
-			errors=true
-		else
-			# output the imgur URL
-			printf '%s    (Delete Hash = %s)\n' "$url" $(get_value_from_response "$response" "deletehash")
+# Prints, notifies & logs given informational message.
+#
+# @param {opt}      -N   OPTIONAL; don't open notification window.
+# @param {opt}      -L   OPTIONAL; don't log the message.
+# @param {string}   msg   message.
+#
+# @returns {void}
+info() {
+    local opt no_notif no_log msg OPTIND
+
+    while getopts "NL" opt; do
+        case "$opt" in
+           N) no_notif=1
+                ;;
+           L) no_log=1
+                ;;
+           *) usage
+              fail
+                ;;
+        esac
+    done
+    shift $((OPTIND-1))
+
+    msg="$1"
+
+    >&2 echo -e "${COLORS[YELLOW]}${SELF}:${COLORS[OFF]} ${msg:-"--info lvl message placeholder--"}"
+
+    if [[ "$no_notif" -ne 1 ]]; then
+        notify INFO "$msg"
+    fi
+
+    if [[ "$no_log" -ne 1 ]]; then
+        echo "INFO [$JOB_ID] $msg" >> "$LOG_FILE"
+    fi
+}
+
+
+# Checks if all required dependiencies are met.
+#
+# @returns {bool}  true, if all required dependencies are installed; false if
+#                  _any_ is missing.
+check_dependencies() {
+    local prog fail
 
-			# append the URL to a string so we can put them all on the clipboard later
-			clip="$clip$url
-		"
-		fi
-	done
+    fail=0
+    for prog in curl jq; do
+        command -v "$prog" >/dev/null || { err -L "[$prog] not installed"; fail=1; }
+    done
 
-	# copy the URLs to the clipboard
-	copy_urls_to_clipboard "$clip"
+    return $fail
 }
 
 
+check_oauth2_client_secrets() {
+    if [[ -z "$CLIENT_ID" || -z "$CLIENT_SECRET" || -z "$REFRESH_TOKEN" ]]; then
+        info -L "Please follow https://apidocs.imgur.com/#authorization-and-oauth in"
+        info -L "order to register your app for [client_id, client_secret, refresh_token]."
+        info -L "When acquired, save those values in [$CREDENTIALS_FILE]"
+        info -L "as CLIENT_ID, CLIENT_SECRET, REFRESH_TOKEN"
+        fail -L
+    fi
+}
+
 
+# Loads required authentication configuration required for authenticated api calls.
 #
-# Deletes the images corresponding to the supplied deletehashes.
+# @returns {void}
+load_access_token() {
+    local expired preemptive_refresh_time
+
+    TOKEN_EXPIRE_TIME=0  # set default, in case it can't be found in $CREDENTIALS_FILE yet
+
+    [[ -s "$CREDENTIALS_FILE" ]] && source "$CREDENTIALS_FILE"
+    check_oauth2_client_secrets
```

Comment:
Initially, there's no `REFRESH_TOKEN`, so this will fail on the first use.

Reply:
But it printed the instructions in console, correct?

Would prefer for uses to do the initial setup manually for simplicity's sake.

Reply:
Wait, what's the non-manual way?

Anyway, yes, it does print the message, but Imgur app registration UI gives you only `CLIENT_ID` and `CLIENT_SECRET`, there's no `REFRESH_TOKEN`, and this check fails when it doesn't find it in the config.

Reply:
Guess I need to re-visit the app registration. Last time I followed that link printed in console, I did end up with refresh token as well. Might've taken additional steps and later forgot about it.

Reply:
[This form](https://api.imgur.com/oauth2/addclient), right?

Reply:
Just in case, `imgur-screenshot` works without a pre-existing refresh token.

Reply:
It's not that easy.
Basically yes, that form gives us `client_id` & `client_secret`, which then need to be used over auth api to receive `refresh_token` (via https://api.imgur.com/oauth2/authorize?client_id=YOUR_CLIENT_ID&response_type=token)

Note imgur-screenshot uses PIN auth, which is deprecated.

Added a bit more descriptive err message, but bottom line is users need to get it sorted manually beforehand. If it can be improved, I'm all ears.

Reply:
Dang, didn't know about the PIN auth deprecation.

Maybe at least make the process interactive? First, ask users to paste in their app `client_id`, and then generate a link they need to follow, and then ask them to copy the redirect URL and paste it back into the app, so then `imgurbash2` can parse it programmatically. More work though.

Reply:
Not a bad idea. i'll give it a go.